src/java/com/puppetlabs/http/client/impl/Metrics.java

```diff
@@ -0,0 +1,158 @@
+package com.puppetlabs.http.client.impl;
+
+import com.codahale.metrics.MetricRegistry;
+import com.codahale.metrics.Timer;
+import org.apache.http.HttpRequest;
+import org.apache.http.RequestLine;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.concurrent.TimeUnit;
+
+public class Metrics {
+    public static final String METRIC_NAMESPACE = "puppetlabs.http-client.experimental";
+    public static final String URL_NAMESPACE = "with-url";
+    public static final String ID_NAMESPACE = "with-metric-id";
+    public static final String BYTES_READ_STRING = "bytes-read";
+    public enum MetricType { BYTES_READ; }
+
+    public static String metricTypeString(Metrics.MetricType metricType) {
+        // currently this is the only metric type we have; in the future when
+        // there are multiple types this will do something more useful
+        return BYTES_READ_STRING;
+    }
+
+    private static ArrayList<Timer.Context> startBytesReadMetricIdTimers(MetricRegistry registry,
+                                                                         String[] metricId) {
+        ArrayList<Timer.Context> timers = new ArrayList<>();
+        for (int i = 0; i < metricId.length; i++) {
+            ArrayList<String> currentId = new ArrayList<>();
+            for (int j = 0; j <= i; j++) {
+                currentId.add(metricId[j]);
+            }
+            currentId.add(0, ID_NAMESPACE);
```

Comment:
If you put this line before the preceding for loop, could this just be `currentId.add(ID_NAMESPACE)`?